File: language/runtime/android/kotlin/src/main/kotlin/dev/destack/runtime/android/bridge/location/LocationAbi.kt
package dev.destack.runtime.android.bridge.location

import dev.destack.runtime.android.bridge.RuntimeAbiStatus
import dev.destack.runtime.android.core.HostSessionHandle
import dev.destack.runtime.android.module.location.RuntimeHostLocationSample

/**
 * The low-level location ingress ABI for one Android runtime bridge.
 */
public interface LocationAbi {
    /**
     * Deliver one location sample into one runtime session.
     */
    public fun notifyLocationSample(
        sessionHandle: HostSessionHandle,
        watchId: String,
        sample: RuntimeHostLocationSample,
    ): RuntimeAbiStatus
}

File: language/runtime/apple/ios/Sources/RuntimeHostIOS/Bridge/Location/LocationAbi.swift
import RuntimeHostAppleBridgeC
import RuntimeHostAppleCore

/// One C callback for one immediate location-services request.
typealias LocationServicesEnabledCallback =
  @convention(c) (UInt64, UnsafeMutablePointer<Bool>?) -> UInt32

/// One C callback for one immediate last-known-location request.
typealias LocationLastKnownCallback =
  @convention(c) (UInt64, UnsafeMutablePointer<DestackRustLocationSample>?) -> UInt32

/// One C callback for one location-watch open request.
typealias LocationWatchOpenCallback =
  @convention(c) (UInt64, DestackRustStringRef, DestackRustLocationWatchOptions) -> UInt32

/// One C callback for one location-watch close request.
typealias LocationWatchCloseCallback =
  @convention(c) (UInt64, DestackRustStringRef) -> UInt32

/// The low-level location ingress ABI for one iOS runtime bridge.
protocol LocationAbi: Sendable {
  /// Deliver one location sample into one runtime session.
  func notifyLocationSample(
    sessionHandle: HostSessionHandle,
    watchID: String,
    sample: RuntimeHostLocationSample
  ) -> RuntimeAbiStatus
}
